tests/unittest/test_helpers.py

```diff
@@ -166,6 +173,981 @@ def test_merge_configurations(self):
+        output_unlabel_rolebinding = ""
+        cmd_label_new_service_account = f"{test_id}/kubectl --kubeconfig {kubeconfig} --namespace {namespace} --context {context} label serviceaccount {username} app.kubernetes.io/managed-by=spark-client app.kubernetes.io/spark-client-primary=1"
+        output_label_service_account = ""
+        cmd_label_new_rolebinding = f"{test_id}/kubectl --kubeconfig {kubeconfig} --namespace {namespace} --context {context} label rolebinding {username}-role app.kubernetes.io/managed-by=spark-client app.kubernetes.io/spark-client-primary=1"
+        output_label_new_rolebinding = ""
+
+        values_subprocess = {
+            cmd_retrieve_primary_sa_yaml: output_retrieve_primary_sa_yaml
+        }
+
+        values_os = {
+            cmd_create_service_account: output_create_service_account,
+            cmd_create_role_binding: output_create_role_binding,
+            cmd_unlabel_service_account: output_unlabel_service_account,
+            cmd_unlabel_rolebinding: output_unlabel_rolebinding,
+            cmd_label_new_service_account: output_label_service_account,
+            cmd_label_new_rolebinding: output_label_new_rolebinding,
+        }
+
+        # test logic
+        env_snap = os.environ.get("SNAP")
+        os.environ["SNAP"] = test_id
+
+        defaults = dict()
+        defaults[conf_key] = conf_value
+        helpers.utils.set_up_user(
+            username, namespace, kubeconfig, context, defaults, mark_primary=True
+        )
+
+        if env_snap:
+            os.environ["SNAP"] = env_snap
+
+        mock_os_system.assert_any_call(cmd_create_service_account)
+        mock_os_system.assert_any_call(cmd_create_role_binding)
+
+        mock_subprocess.assert_any_call(cmd_retrieve_primary_sa_yaml, shell=True)
+
+        mock_os_system.assert_any_call(cmd_unlabel_service_account)
+        mock_os_system.assert_any_call(cmd_unlabel_rolebinding)
+        mock_os_system.assert_any_call(cmd_label_new_service_account)
+        mock_os_system.assert_any_call(cmd_label_new_rolebinding)
+
+    @patch("helpers.utils.os.system")
+    @patch("helpers.utils.subprocess.check_output")
+    def test_set_up_user_primary_defined_primary_not_reassigned(
+        self, mock_subprocess, mock_os_system
+    ):
+        # mock logic
+        def side_effect_subprocess(*args, **kwargs):
+            return values_subprocess[args[0]]
+
+        def side_effect_os(*args, **kwargs):
+            return values_os[args[0]]
+
+        mock_os_system.side_effect = side_effect_os
+        mock_os_system.return_value = 0
+
+        mock_subprocess.side_effect = side_effect_subprocess
+
+        test_id = str(uuid.uuid4())
+        username = str(uuid.uuid4())
+        namespace = str(uuid.uuid4())
+        kubeconfig = str(uuid.uuid4())
+        context = str(uuid.uuid4())
+        conf_key = str(uuid.uuid4())
+        conf_value = str(uuid.uuid4())
```

Comment:
Just saying but it does not need refactoring, giving that we will anyway need to refactor unittests, after the structure refactoring we are planning, but many tests have very similar piece of code for mocking. This could in general be factor out 

But as I said, just raising this point when we will restructure unit tests next time. At this point, there is no action required!

Reply:
I also suggest to have to have a module that contains shared functions for the tests.